AEXS/aex-141.md

```diff
@@ -372,52 +431,40 @@ stateful entrypoint check_swap(account: address) : int
 
 ### swapped\(\)
 
-Returns all of the swapped NFTs that are stored in contract state. 
+Returns the map with the swapped NFTs per address.
 
 ```sophia
-stateful entrypoint swapped() : map(address, int)
+stateful entrypoint swapped() : map(address, Set.set(int))
 ```
 
 | return | type |
 | :--- | :--- |
-| swapped | map(address, int) |
+| swapped | map(address, Set.set(int)) |
 
 ## Events
 
-**Swap** - MUST trigger when tokens are swapped using the `swap` function.
+**Burn** - MUST trigger when NFTs are burned using the `burn` function.
 
-The swap event arguments should be as follows: `(account,  value)`
+The burn event arguments should be as follows: `(owner, token_id)`
 
 ```sophia
-Swap(address, int)
+Burn(address, int)
 ```
 
 | parameter | type |
 | :--- | :--- |
-| account| address |
-| value | int |
+| owner | address |
+| token_id | int |
 
-# Receiver contract interface
+**Swap** - MUST trigger for each NFT that is swapped using the `swap` function.
 
-## NFTReceiver
+The swap event arguments should be as follows: `(owner,  token_id)`
```

Comment:
@jyeshe are you sure? we wouldn't emit sth. for the `token_id` then. in case of fungible tokens that's not a problem. but I think in case of NFTs I'd expect to know which specific NFTs have been swapped.

I first thought about a string which includes all IDs concatenated and comma seperated.

what do you think @thepiwo ?

Reply:
Because the swap might burn multiple tokens.

Reply:
Yes but for Burn or Mint we could also have custom entrypoints that perform batch operations where you'd have to emit the event X times.

So current assumption from me would be emitting the Swap event for each NFT.

If we don't care about the NFT id in the event we can leave as is and only use the "balance".

Another option would be concatenation in a string that includes all NFT ids

Reply:
how does ERC 721 do it?

Reply:
I don't think swap exists there. Need to look up.

Reply:
yes, couldn't find sth. like that. so it's up to us to decide how to deal with that event

Reply:
There is batch event on ERC1155 declared like this with `uint256[] _ids`:
`
event TransferBatch(address indexed _operator, address indexed _from, address indexed _to, uint256[] _ids, uint256[] _values);
`
This is not accepted: `datatype event = Swap(address, list(int))` and would go with `Swap(address)` because the indexer (including the Mdw) already knows which tokens the `address` owns to to remove ownership for these swapped tokens. Would follow the idea that data that is inferred tend to be more consistent than redundant/duplicated ones.

Reply:
`datatype event = Swap(address, list(int))` I know, this is why I only see two options in case we wanna include the NFT ids:
- emit for each NFT
OR
- use `string` and concatenate all IDs in a comma separated string

if we don't care about the IDs we can go with `Swap(address)`

Reply:
either we allow complete "following" of nfts via events or not at all, swap without reference to token doesn't make sense

Reply:
agree with you. actually I prefer `Swap(address, int)` like we have right now. the same applies for `Mint(address, int)`

you can have batch-minting but each mint must emit the event individually

Reply:
> either we allow complete "following" of nfts via events or not at all, swap without reference to token doesn't make sense

It informs the burner, that's enough for a indexer. It would make more sense than emitting multiple events for an event that happened only once (a single `swap`).

Reply:
that's a fair point. if you think it's sufficient I am fine with that. but I somehow like to be explicit with `Swap(address, int)` event for each swapped NFT. we already do the same for minting. the key difference is that this is a one-time event 🤔 

actually I have not such a strong opinion. I am fine with both.

Reply:
With current gas price, couldn't emit more than 2300 `Swap(address, int)` events on a method that only emits the event using `gasLimit=5817123` for a `GasLimitMax=5818080` (limited by the SDK)

Reply:
yeah, I thought about the gaslimit anyway in that regards. I think we cannot apply the same swap logic as in AEX-9 as a user could potentially have a lot of NFTs as you mentioned.

I think we even need to provide different swap entrypoints:
- `swap(int)` for a swap of a single NFT
- `swap(Set.set(int))` for a swap of multiple NFTs

and then keep `Swap(address, int)` event

Reply:
another option would be to keep `swap()` without any argument and return a tuple of `(int, int, bool)` which states `(amount_swapped, amount_remaining, finished_swap)`

then we would be flexible in terms of implementation.

still it remains open how to deal with the event. I still prefer `Swap(address, int)` for every single NFT.

opinions again? @jyeshe @thepiwo

Reply:
Two use cases for `swap` that might be a reason to avoid `Swap(address, int)` which is limited to 2300 events:
1. When a collection doesn't perform as expected by creators. The collection is then discontinued by burning all pre-minted tokens using `swap` (without pre-mint there would be no `swap`, just a stop on the contract).
2. When the creators want to decrease the planned supply after launch (instead of burning multiple times after minting).

Besides that multiple `Swap(address, int)` events for a single `swap` call still sound unusual for me.

Regarding the amounts arguments, as an indexer I would rely on the actual owned tokens by the swapper/burner.

Reply:
IMO `Swap` is only intended to be used for "swapping" NFTs to a new contract **OR** to another chain.

for burning NFTs `burn` should be used. **IF** single burns are not sufficient or efficient enough we would have to adjust that extension.

for reasons mentioned above I really like the flexibility we achieve by keeping `swap()` without any argument and return a tuple of `(int, int, bool)` which states `(amount_swapped, amount_remaining, finished_swap)`.

an additional reason for that is that AFTER a swap an address could theoretically receive new tokens and would potentially have to swap again.

keeping `Swap(address, int)` is just consistent in that way. `swap` isn't necessarily a one-time event.

maybe we should clarify the usage of `Swappable Extension`. at least this was my assumption all the time. cc @thepiwo

Reply:
I am following the definition of "Burns the whole balance of Call.caller and and stores the same amount in the swapped map". Based on this, someone else might prefer it than burning multiple times.

Either way, whatever the use case, could it be limited to 2300 event with `Swap(address, int)`? 
For the alternative of `(amount_swapped, amount_remaining, finished_swap)` could you give an example of how it would be used by an indexer? Not mentioning a trader because I believe traders rely on as much info as possible from a collection. It's important to know all tokens that were burned to keep track of both supply and rarity.

Reply:
@jyeshe I was maybe not clear enough regarding the tuple. the tuple `(amount_swapped, amount_remaining, finished_swap)` would be the return value when calling the function for the Caller to immediately know if there are still NFTs remaining and if `swap` would have to be called again. but I think we could simplify the tuple and just return `(amount_swapped, amount_remaining)`.

my assumption right now is that we might not be able to swap all the NFTs **at once** because there has to be lookups and we might run out of gas if somebody owns a huge amount of NFTs in a collection.

the event would still be `Swap(address, int)` for each single NFT.

let's assume `swap` entrypoint can handle ~ 2000 NFTs at once and the account owns 3000 NFTs. then you'd have to call `swap` twice:
1. `swap` -> returns `(2000, 1000)`
2. `swap` -> returns `(3000, 0)`

and according to the event `Swap(address, int)` -> it would be called 3000 times in total (for each NFT) 😅

Reply:
Yes, swap in AEX-9 was prepared to be able to swap to native tokens, another standard or similar.

For swap it would be crucial to also provide an example on how to swap in reality and a testcase for this, so we can be sure it works as expected.

Reply:
@thepiwo what's your opinion about the event and the proposal with returning the tuple with `(amount_swapped, amount_remaining)` ?

Reply:
I don't think amount_remaining is needed, also who cares about "amount of nfts" if I track nfts I need to know events for every single token otherwise its useless

Reply:
> @jyeshe I was maybe not clear enough regarding the tuple. the tuple `(amount_swapped, amount_remaining, finished_swap)` would be the return value when calling the function for the Caller to immediately know if there are still NFTs remaining and if `swap` would have to be called again. but I think we could simplify the tuple and just return `(amount_swapped, amount_remaining)`.
> 
> my assumption right now is that we might not be able to swap all the NFTs **at once** because there has to be lookups and we might run out of gas if somebody owns a huge amount of NFTs in a collection.
> 
> the event would still be `Swap(address, int)` for each single NFT.
> 
> let's assume `swap` entrypoint can handle ~ 2000 NFTs at once and the account owns 3000 NFTs. then you'd have to call `swap` twice:
> 
>     1. `swap` -> returns `(2000, 1000)`
> 
>     2. `swap` -> returns `(3000, 0)`
> 
> 
> and according to the event `Swap(address, int)` -> it would be called 3000 times in total (for each NFT) sweat_smile

Feel free to choose the best flavor for you, having the `address` on the event is what matters for an indexer and for some listener to call `check_swap`. Having it we make it work anyway (:

Reply:
> I don't think amount_remaining is needed

generally agree as we could check with `balance(address)` if needed. do we agree there could be scenarios where we would run out of gas, so we cannot assume that ALL NFTs can be swapped at once?

in this case I think `amount_remaining` makes sense in that regards so that calling `balance(address)` can be avoided. 

`amount_swapped` can definitely wiped out, agree

Reply:
suggesting to return `bool` now to indicate a complete swap

Reply:
I will perform some tests swapping large amount of tokens. I would recommend every early stage NFT contract to implement this as there might be a need to switch to new implementations later on

Reply:
removed swappable extension